### ISSUE TITLE
JDK 8/11/17/18 - incorrect result from JIT compilation due to NegativeArraySizeException not thrown

### ISSUE THREAD

**Issue by JeffersonYu1** [eclipse-openj9/openj9#17139]
# JDK 8/11/17/18 - incorrect result from JIT compilation due to NegativeArraySizeException not thrown

### Java version
$ java -version
```
openjdk version "17.0.6" 2023-01-17
IBM Semeru Runtime Open Edition 17.0.6.0 (build 17.0.6+10)
Eclipse OpenJ9 VM 17.0.6.0 (build openj9-0.36.0, JRE 17 Linux amd64-64-Bit Compressed References 20230117_397 (JIT enabled, AOT enabled)
OpenJ9   - e68fb241f
OMR      - f491bbf6f
JCL      - 927b34f84c8 based on jdk-17.0.6+10)
```

### Operating system details
$ cat /etc/*release
```
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=20.04
DISTRIB_CODENAME=focal
DISTRIB_DESCRIPTION="Ubuntu 20.04.3 LTS"
NAME="Ubuntu"
VERSION="20.04.3 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.3 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
```

$ uname -a
```
Linux luzhou 5.4.0-86-generic #97-Ubuntu SMP Fri Sep 17 19:19:40 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

## Description

Incorrect result is given when running the following program. The bug affects 8u362, 11.0.18.0, 17.0.6.0, 18.0.2.1.

## Steps to reproduce
The following steps shows how to reproduce the bug on JDK 17.0.6.0 in a Ubuntu Linux environment.

### Compile
$ javac C.java

### Execute
$ java C
-6670   // nondeterministic value

$ java -Xnojit C
0       // correct value

### Source code for an executable test case
// C.java
```
public class C {
    public static int m(float[] array) {
        // should throw NegativeArraySizeException
        float[] subarray = new float[array.length - 1];
        return subarray.length;
    }

    public static void main(String[] args) {
        int sum = 0;
        for (int i = 0; i < 10000; ++i) {
            try {
                sum += m(new float[]{});
            } catch (NegativeArraySizeException ignored) {
            }
        }
        System.out.println(sum); // expected 0
    }
}
```

### Workaround
Disable JIT.
$ java -Xnojit C

**Comment by pshipton** [comment]
@0xdaryl @hzongaro pls take a look.

**Comment by hzongaro** [comment]
It looks like this might be a problem in Dead Trees Elimination.  Before that optimization, the IL for `C.m` looks like this, in part:

```
n1n       BBStart <block_2> (freq 8001)                                                       [0x7f32f59395b0] bci=[-1,0,4] rc=0 vc=158 vn=- li=- udi=- nc=1
n50n        GlRegDeps                                                                         [0x7f32f593a500] bci=[-1,0,4] rc=1 vc=158 vn=- li=- udi=- nc=1
n51n          aRegLoad eax   <parm 0 [F>[#417  Parm] [flags 0x40000107 0x0 ] (SeenRealReference )  [0x7f32f593a550] bci=[-1,0,4] rc=2 vc=158 vn=- li=- udi=- nc=0 flg=0x8000
n5n       NULLCHK on n51n [#32]                                                               [0x7f32f59396f0] bci=[-1,1,4] rc=0 vc=158 vn=- li=- udi=- nc=1
n4n         arraylength (stride 4) (X>=0 cannotOverflow )                                     [0x7f32f59396a0] bci=[-1,1,4] rc=2 vc=158 vn=- li=- udi=- nc=1 flg=0x1100
n51n          ==>aRegLoad
n10n      treetop                                                                             [0x7f32f5939880] bci=[-1,4,4] rc=0 vc=158 vn=- li=- udi=- nc=1
n9n         newarray  jitNewArray[#102  helper Method] [flags 0x400 0x0 ] (highWordZero Unsigned X!=0 allocationCanBeRemoved )  [0x7f32f5939830] bci=[-1,4,4] rc=2 vc=158 vn=- li=- udi=- nc=2 flg=0x4004
n7n           iadd (cannotOverflow )                                                          [0x7f32f5939790] bci=[-1,3,4] rc=2 vc=158 vn=- li=- udi=- nc=2 flg=0x1000
n4n             ==>arraylength
n6n             iconst -1 (X!=0 X<=0 )                                                        [0x7f32f5939740] bci=[-1,2,4] rc=1 vc=158 vn=- li=- udi=- nc=0 flg=0x204
n8n           iconst 6   ; array type is float (X!=0 X>=0 )                                   [0x7f32f59397e0] bci=[-1,4,4] rc=1 vc=158 vn=- li=- udi=- nc=0 flg=0x104
n11n      treetop                                                                             [0x7f32f59398d0] bci=[-1,6,4] rc=0 vc=158 vn=- li=- udi=1 nc=1
n9n         ==>newarray
n15n      ireturn                                                                             [0x7f32f5939a10] bci=[-1,9,5] rc=0 vc=158 vn=- li=- udi=- nc=1
n7n         ==>iadd
n2n       BBEnd </block_2> =====                                                              [0x7f32f5939600] bci=[-1,9,5] rc=0 vc=158 vn=- li=- udi=- nc=0
```

Dead Trees elimination removes the `jitNewArray` operation which was expected to throw a `NegativeArraySizeException`:

```
Performing 92: deadTreesElimination
[    24]  92.1    O^O DEAD TREES ELIMINATION: Remove trivial dead tree: 00007F32F59398D0
[    25]  92.2    O^O DEAD TREES ELIMINATION: Remove tree : [00007F32F5939880] ([00007F32F5939830] = newarray)
[    26]  92.3    O^O DEAD TREES ELIMINATION: Remove tree : [00007F32F5B25190] ([00007F32F5939790] = iadd)
  ...
n1n       BBStart <block_2> (freq 8001)                                                       [0x7f32f59395b0] bci=[-1,0,4] rc=0 vc=161 vn=- li=-1 udi=- nc=1
n50n        GlRegDeps                                                                         [0x7f32f593a500] bci=[-1,0,4] rc=1 vc=161 vn=- li=- udi=- nc=1
n51n          aRegLoad eax   <parm 0 [F>[#417  Parm] [flags 0x40000107 0x0 ] (SeenRealReference )  [0x7f32f593a550] bci=[-1,0,4] rc=2 vc=161 vn=- li=- udi=- nc=0 flg=0x8000
n5n       NULLCHK on n51n [#32]                                                               [0x7f32f59396f0] bci=[-1,1,4] rc=0 vc=161 vn=- li=-1 udi=- nc=1
n4n         arraylength (stride 4) (X>=0 cannotOverflow )                                     [0x7f32f59396a0] bci=[-1,1,4] rc=2 vc=161 vn=- li=- udi=- nc=1 flg=0x1100
n51n          ==>aRegLoad
n15n      ireturn                                                                             [0x7f32f5939a10] bci=[-1,9,5] rc=0 vc=161 vn=- li=-1 udi=- nc=1
n7n         iadd (cannotOverflow )                                                            [0x7f32f5939790] bci=[-1,3,4] rc=1 vc=161 vn=- li=- udi=- nc=2 flg=0x1000
n4n           ==>arraylength
n6n           iconst -1 (X!=0 X<=0 )                                                          [0x7f32f5939740] bci=[-1,2,4] rc=1 vc=161 vn=- li=- udi=- nc=0 flg=0x204
n2n       BBEnd </block_2> =====                                                              [0x7f32f5939600] bci=[-1,9,5] rc=0 vc=161 vn=- li=-1 udi=- nc=0
```

**Comment by hzongaro** [comment]
I took another quick look, and I think it might be that Dead Trees Elimination thinks it's safe to remove the `newarray` operation because the node has the `allocationCanBeRemoved` flag set.  That comes to be during an earlier pass of Global Value Propagation.

**Comment by hzongaro** [comment]
Fixed by OMR pull request eclipse/omr#6952
